core/src/main/java/org/mule/runtime/core/privileged/routing/RoutingResult.java

```diff
@@ -29,18 +33,44 @@ public final class RoutingResult {
 
   private final Map<String, Message> successfulRoutesResultMap;
   private final Map<String, Error> failedRoutesErrorMap;
+  private Map<String, Pair<Error, EventProcessingException>> failedRoutesErrorWithExceptionMap;
 
   public RoutingResult(Map<String, Message> successfulRoutesResultMap, Map<String, Error> failedRoutesErrorMap) {
     this.successfulRoutesResultMap = unmodifiableMap(successfulRoutesResultMap);
     this.failedRoutesErrorMap = unmodifiableMap(failedRoutesErrorMap);
+    this.failedRoutesErrorWithExceptionMap = emptyMap();
+  }
+
+  public static RoutingResult routingResultWithException(Map<String, Message> successfulRoutesResultMap,
+                                                         Map<String, Pair<Error, EventProcessingException>> failedRoutesErrorWithExceptionMap) {
+    RoutingResult routingResult = new RoutingResult(successfulRoutesResultMap, emptyMap());
+    routingResult.setFailedRoutesErrorWithExceptionMap(failedRoutesErrorWithExceptionMap);
+    return routingResult;
+  }
+
+  private void setFailedRoutesErrorWithExceptionMap(Map<String, Pair<Error, EventProcessingException>> failedRoutesErrorWithExceptionMap) {
+    this.failedRoutesErrorWithExceptionMap = failedRoutesErrorWithExceptionMap;
   }
 
   public Map<String, Message> getResults() {
     return successfulRoutesResultMap;
   }
 
   public Map<String, Error> getFailures() {
```

Comment:
Code will look a lot better if you start with if (!failedRoutesErrorWithExceptionMap.isEmpty()) { ... } else {return failedRoutesErrorMap;}